File: 剑指offer/18.删除链表的节点.py
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None


class Solution:
    def deleteNode(self, head: ListNode, val: int) -> ListNode:
        if not head:
            return None

        new_head = ListNode(0)
        new_head.next = head

        p = new_head
        q = p.next

        while q:
            if q.val == val:
                p.next = q.next
                break
            else:
                p = q
                q = p.next

        return new_head.next

File: 剑指offer/18.删除链表的节点.cpp
#include <iostream>
using namespace std;

struct ListNode {
    int val;
    ListNode* next;
    ListNode(int x)
        : val(x)
        , next(NULL)
    {
    }
};

class Solution {
public:
    ListNode* deleteNode(ListNode* head, int val)
    {
        ListNode* new_head = new ListNode();
        new_head->next = head;

        ListNode* pre = new_head;
        while (pre->next != NULL) {
            if (pre->next->val == val) {
                pre->next = pre->next->next;
                break;
            } else {
                pre = pre->next;
            }
        }

        return new_head->next;
    }
};